website/src/_posts/2017-02-13-stateful-processing.md

```diff
@@ -499,31 +540,59 @@ new DoFn<KV<UserId, Event>, KV<UserId, Prediction>>() {
 ```
 
 ```py
-# State and timers are not yet supported in Beam's Python SDK.
-# Watch this space!
+class ModelStatefulFn(beam.DoFn):
+
+  PREVIOUS_PREDICTION = BagStateSpec('previous_pred_state', PredictionCoder())
+  MODEL_STATE = CombiningValueStateSpec('model_state',
+                                        ModelCoder(),
+                                        ModelFromEventsFn())
+
+  def process(self,
+              user_event,
+              previous_pred_state=beam.DoFn.StateParam(PREVIOUS_PREDICTION),
+              model_state=beam.DoFn.StateParam(MODEL_STATE)):
+    user = user_event[0]
+    event = user_event[1]
+    model = model_state.read()
+    previous_prediction = previous_pred_state.read()
+
+    new_prediction = model.prediction(event)
+    model_state.add(event)
+
+    if (previous_prediction is None
+        or self.should_output_prediction(
+            previous_prediction, new_prediction)):
+      previous_pred_state.clear()
+      previous_pred_state.add(new_prediction)
+      yield (user, new_prediction)
 ```
 
 Let's walk through it,
 
- - You have two state cells declared, `@StateId("model")` to hold the current
-   state of the model for a user and `@StateId("previousPrediction")` to hold
+ - You have two state cells declared,
+   `@StateId("model")/StateParam(MODEL_STATE)` to hold the current
+   state of the model for a user and
+   `@StateId("previousPrediction")/StateParam(PREVIOUS_PREDICTION)` to hold
    the prediction output previously.
  - Access to the two state cells by annotation in the `@ProcessElement` method
-   is as before.
- - You read the current model via `modelState.read()`. Because state is also
-   per-key-and-window, this is a model just for the UserId of the Event
-   currently being processed.
+   is as before, and in Python by using an argument placeholder.
```

Comment:
You know, at this point it is starting to see like it should be documentation rather than editing the blog post. I say this because blog posts are dated and it is not great to go back and rewrite them like this. I know it gets a lot of hits, but probably we just need docs and they can receive all the hits.

Reply:
I agree, we shouldn't be rewriting the prose section of the blog post, but I still think that adding the Python code to the blog post has value.

Reply:
I believe you can add span around the Java/Python code, e.g. `<span language=py>Python code</span>` which will only show the selected language. This way the prose is not distorted with two languages here.

Reply:
I agree we need documentation, but I also think it's worth updating this post (and not holding that up further). I would exclude the prose changes except the ", and in Python by using an argument placeholder" phrase which I think may not be as obvious as the others. (We could do fancy things with language-specific spans, but I would also be fine letting the prose remain, with that one addition, as it was.)

Reply:
+1 I also think it is worth getting this in as a reference point before we add further documentation.